pip/req/req_set.py

```diff
@@ -238,6 +238,15 @@ def add_requirement(self, install_req, parent_req_name=None):
                 existing_req = None
             if (parent_req_name is None and existing_req and not
                     existing_req.constraint):
+                if existing_req.name == install_req.name:
+                    # If the extras do not match, then union them together to
+                    # support install bar bar[foo]. If they do match, the
+                    # InstallationError will be raised as per normal.
+                    if existing_req.extras != install_req.extras:
+                        existing_req.extras = tuple(
+                            set(existing_req.extras).union(
+                                set(install_req.extras)))
+                        return [existing_req]
```

Comment:
It seems as though this code should be located elsewhere.

If the requirement has not been switched from a constraint to a req, the req on the command line is silently dropped.

" and not existing_req.constraint" is killing this code.